File: quality_of_service_demo/rclcpp/src/deadline.cpp
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>

#include "rcutils/cmdline_parser.h"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/executors/single_threaded_executor.hpp"

#include "std_msgs/msg/string.hpp"
#include "std_msgs/msg/bool.hpp"

#include "quality_of_service_demo/common_nodes.hpp"

static const char * OPTION_PUBLISH_FOR = "--publish-for";
static const size_t DEFAULT_PUBLISH_FOR = 5000;
static const char * OPTION_PAUSE_FOR = "--pause-for";
static const size_t DEFAULT_PAUSE_FOR = 1000;

void print_usage()
{
  printf("Usage for deadline:\n");
  printf(
    "deadline "
    "deadline_duration "
    "[%s publish_for] "
    "[%s pause_for] "
    "[-h]\n",
    OPTION_PUBLISH_FOR,
    OPTION_PAUSE_FOR);
  printf("required arguments:\n");
  printf(
    "deadline_duration: "
    "Duration in positive integer milliseconds of the Deadline QoS setting.\n");
  printf("optional arguments:\n");
  printf("-h : Print this help message.\n");
  printf(
    "%s publish_for : "
    "Duration to publish (in positive integer milliseconds) until pausing the talker. "
    "Defaults to %zu.\n",
    OPTION_PUBLISH_FOR, DEFAULT_PUBLISH_FOR);
  printf(
    "%s pause_for : "
    "Duration to pause (in positive integer milliseconds) before starting to publish again. "
    "Defaults to %zu.\n",
    OPTION_PAUSE_FOR, DEFAULT_PAUSE_FOR);
}

int main(int argc, char * argv[])
{
  // Force flush of the stdout buffer.
  setvbuf(stdout, NULL, _IONBF, BUFSIZ);

  // Argument count and usage
  if (argc < 2 || rcutils_cli_option_exist(argv, argv + argc, "-h")) {
    print_usage();
    return 0;
  }

  // Configuration variables
  std::chrono::milliseconds deadline_duration(std::stoul(argv[1]));
  std::chrono::milliseconds period_pause_talker(DEFAULT_PUBLISH_FOR);
  std::chrono::milliseconds duration_pause_talker(DEFAULT_PAUSE_FOR);
  std::string topic("qos_deadline_chatter");

  // Optional argument parsing
  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_PUBLISH_FOR)) {
    period_pause_talker = std::chrono::milliseconds(
      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_PUBLISH_FOR)));
  }
  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_PAUSE_FOR)) {
    duration_pause_talker = std::chrono::milliseconds(
      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_PAUSE_FOR)));
  }

  // Initialization and configuration
  rclcpp::init(argc, argv);
  rclcpp::executors::SingleThreadedExecutor executor;

  rclcpp::QoS qos_profile(10);
  qos_profile.deadline(deadline_duration);

  auto talker = std::make_shared<Talker>(qos_profile, topic);
  talker->get_options().event_callbacks.deadline_callback =
    [node = talker.get()](rclcpp::QOSDeadlineOfferedInfo & event) -> void
    {
      RCLCPP_INFO(
        node->get_logger(), "Offered deadline missed - total %d delta %d",
        event.total_count, event.total_count_change);
    };

  auto listener = std::make_shared<Listener>(qos_profile, topic);
  listener->get_options().event_callbacks.deadline_callback =
    [node = listener.get()](rclcpp::QOSDeadlineRequestedInfo & event) -> void
    {
      RCLCPP_INFO(
        node->get_logger(), "Requested deadline missed - total %d delta %d",
        event.total_count, event.total_count_change);
    };

  talker->initialize();
  listener->initialize();

  auto pause_timer = talker->create_wall_timer(
    period_pause_talker,
    [talker, duration_pause_talker]() -> void {
      talker->pause_publish_for(duration_pause_talker);
    });

  // Execution
  executor.add_node(talker);
  executor.add_node(listener);
  executor.spin();

  // Cleanup
  rclcpp::shutdown();
  return 0;
}

File: quality_of_service_demo/rclpy/quality_of_service_demo_py/deadline.py
# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import argparse
import sys

from quality_of_service_demo_py.common_nodes import Listener
from quality_of_service_demo_py.common_nodes import Talker

import rclpy
from rclpy.duration import Duration
from rclpy.executors import ExternalShutdownException
from rclpy.executors import SingleThreadedExecutor
from rclpy.logging import get_logger
from rclpy.qos import QoSProfile
from rclpy.qos_event import PublisherEventCallbacks
from rclpy.qos_event import SubscriptionEventCallbacks


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'deadline', type=int,
        help='Duration in positive integer milliseconds of the Deadline QoS setting.')
    parser.add_argument(
        '--publish-for', type=int, default=5000,
        help='Duration in positive integer milliseconds to publish until pausing the talker.')
    parser.add_argument(
        '--pause-for', type=int, default=1000,
        help='Duration in positive integer milliseconds to pause the talker before beginning '
             'to publish again.')
    return parser.parse_args()


def main(args=None):
    parsed_args = parse_args()
    rclpy.init(args=args)

    topic = 'qos_deadline_chatter'
    deadline = Duration(seconds=parsed_args.deadline / 1000.0)

    qos_profile = QoSProfile(
        depth=10,
        deadline=deadline)

    subscription_callbacks = SubscriptionEventCallbacks(
        deadline=lambda event: get_logger('Listener').info(str(event)))
    listener = Listener(topic, qos_profile, event_callbacks=subscription_callbacks)

    publisher_callbacks = PublisherEventCallbacks(
        deadline=lambda event: get_logger('Talker').info(str(event)))
    talker = Talker(topic, qos_profile, event_callbacks=publisher_callbacks)

    publish_for_seconds = parsed_args.publish_for / 1000.0
    pause_for_seconds = parsed_args.pause_for / 1000.0
    pause_timer = talker.create_timer(  # noqa: F841
        publish_for_seconds,
        lambda: talker.pause_for(pause_for_seconds))

    executor = SingleThreadedExecutor()
    executor.add_node(listener)
    executor.add_node(talker)
    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        sys.exit(1)
    finally:
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
